# Web Mining and Applied NLP (44-620)

## Requests, JSON, and NLP

### Student Name: Benjamin Robin

### GitHub URL: https://github.com/BenjaminRobin22/json-sentiment 

Perform the tasks described in the Markdown cells below.  When you have completed the assignment make sure your code cells have all been run (and have output beneath them) and ensure you have committed and pushed ALL of your changes to your assignment repository.

Make sure you have [installed spaCy and its pipeline](https://spacy.io/usage#quickstart) and [spaCyTextBlob](https://spacy.io/universe/project/spacy-textblob)

Every question that requires you to write code will have a code cell underneath it; you may either write your entire solution in that cell or write it in a python file (`.py`), then import and run the appropriate code to answer the question.

This assignment requires that you write additional files (either JSON or pickle files); make sure to submit those files in your repository as well.

In [1]:
# Create and activate a Python virtual environment. 
# Before starting the project, try all these imports FIRST
# Address any errors you get running this code cell 
# by installing the necessary packages into your active Python environment.
# Try to resolve issues using your materials and the web.
# If that doesn't work, ask for help in the discussion forums.
# You can't complete the exercises until you import these - start early! 
# We also import json and pickle (included in the Python Standard Library).

import json
import pickle

import requests
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

print('All prereqs installed.')
!pip list

All prereqs installed.
Package                   Version
------------------------- ---------------
annotated-types           0.6.0
anyio                     4.3.0
argon2-cffi               23.1.0
argon2-cffi-bindings      21.2.0
arrow                     1.3.0
asttokens                 2.4.1
async-lru                 2.0.4
attrs                     23.2.0
Babel                     2.14.0
beautifulsoup4            4.12.3
bleach                    6.1.0
blis                      0.7.11
catalogue                 2.0.10
certifi                   2024.2.2
cffi                      1.16.0
charset-normalizer        3.3.2
click                     8.1.7
cloudpathlib              0.16.0
colorama                  0.4.6
comm                      0.2.1
confection                0.1.4
contourpy                 1.2.0
cycler                    0.12.1
cymem                     2.0.8
debugpy                   1.8.1
decorator                 5.1.1
defusedxml                0.7.1
executing               

1. The following code accesses the [lyrics.ovh](https://lyricsovh.docs.apiary.io/#reference/0/lyrics-of-a-song/search) public api, searches for the lyrics of a song, and stores it in a dictionary object.  Write the resulting json to a file (either a JSON file or a pickle file; you choose). You will read in the contents of this file for future questions so we do not need to frequently access the API.

In [1]:
import requests
import json
import pickle
# Request for the API 
response = requests.get('https://api.lyrics.ovh/v1/They Might Be Giants/Birdhouse in your soul')
result = response.json()

if response.status_code == 200:
    # Load the JSON data
    result = response.json()
    
    # Save the JSON data to a JSON file
    with open('lyrics.json', 'w') as json_file:
        json.dump(result, json_file, indent=4)
        
    print("Data saved successfully.")
else:
    print("Failed return data.")

Data saved successfully.


In [30]:
import requests
import json

AUTHOR='Edgar Allan Poe'
POEM = 'A Dream Within A Dream'

#only certain poets and titles are available
#to see the available poets, go to (in a web browser)
# https://poetrydb.org/author
#To see which poems that author has available, go to 
# https://poetrydb.org/author/AUTHOR NAME
# e.g.: https://poetrydb.org/author/Edgar Allan Poe
#The spaces will get handled by your web browser

# A cool pythonism (introduced in Python 3): f strings
# https://docs.python.org/3/tutorial/inputoutput.html#tut-f-strings
URL = f'https://poetrydb.org/author,title/{AUTHOR};{POEM}'
result = json.loads(requests.get(URL).text)
poem = '\n'.join(result[0]['lines'])

with open('poem.json', 'w') as file:
    json.dump(result, file)

I was able to get the orignal song lyrics to work. I also imported the alternative expamle as well as I am not sure if the original lyrics would work further on in the assignment. 

2. Read in the contents of your file.  Print the lyrics of the song (not the entire dictionary!) and use spaCyTextBlob to perform sentiment analysis on the lyrics.  Print the polarity score of the sentiment analysis.  Given that the range of the polarity score is `[-1.0,1.0]` which corresponds to how positive or negative the text in question is, do you think the lyrics have a more positive or negative connotaion?  Answer this question in a comment in your code cell.

In [29]:
import requests
import json
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

# Load spaCy language model
nlp = spacy.load('en_core_web_sm')
nlp.add_pipe('spacytextblob')

# Read the contents of the file
with open('lyrics.json', 'r') as json_file:
    results = json.load(json_file)

# Check if the JSON data has the key 'lyrics'
if 'lyrics' in results:
    # Extract the lyrics
    song = results['lyrics']
    
    # Process the lyrics using spaCy
    doc = nlp(song)

    # Perform sentiment analysis
    sentiment_score = doc._.polarity

    # Print the lyrics
    print("Lyrics of the song:")
    print(song)

    # Print the polarity score of the sentiment analysis
    print("Polarity score of sentiment analysis:", sentiment_score)
else:
    print("No lyrics found in the result.")


Lyrics of the song:
Paroles de la chanson Birdhouse In Your Soul par They Might Be Giants
I'm Your Only Friend

I'm Not Your Only Friend

But I'm A Little Glowing Friend

But Really I'm Not Actually Your Friend

But I Am

Blue Canary In The Outlet By The Light Switch

Who Watches Over You

Make A Little Birdhouse In Your Soul

Not To Put Too Fine A Point On It

Say I'm The Only Bee In Your Bonnet

Make A Little Birdhouse In Your Soul

I Have A Secret To Tell

From My Electrical Well

It's A Simple Message And I'm Leaving Out The Whistles And Bells

So The Room Must Listen To Me


Filibuster Vigilantly

My Name Is Blue Canary One Note Spelled Lite

My Story's Infinite

Like The Longines Symphonette It Doesn't Rest

Blue Canary In The Outlet By The Light Switch

Who Watches Over You

Make A Little Birdhouse In Your Soul

Not To Put Too Fine A Point On It

Say I'm The Only Bee In Your Bonnet

Make A Little Birdhouse In Your Soul

I'm Your Only Friend

I'm Not Your Only Friend

But I'm A L

The polarity score shows a very neutral sentiment less than .1 positive. I would have a hard time ranking it myself. For one thing the singer seems to be comming from a dark place or at least a sad one. This started to brighten as the song progresses with the blue canary and other objects. I would tend to agree with the slight positive ranking though I am very bad personally at interpreting songs. 

3. Write a function that takes an artist, song, and filename, accesses the lyrics.ovh api to get the song lyrics, and writes the results to the specified filename.  Test this function by getting the lyrics to any four songs of your choice and storing them in different files.

In [4]:
import requests
import json
# definining attributes 
def write_lyrics_to_json(artist, song, filename):
    # Construct the API URL
    url = f"https://api.lyrics.ovh/v1/{artist}/{song}"

    # Send a GET request to the API
    response = requests.get(url)

    # Check if the request was successful with HTTP 200 typically indicated a successful response according to my research. 
    if response.status_code == 200:
        # Extract the JSON data from the response
        data = response.json()

        # Check if lyrics are available
        if 'lyrics' in data:
            # Write the JSON data to the specified file with 'w' being write mode 
            with open(filename, 'w') as file:
            # Making the file more readable as recomended when creating json files. 
                json.dump(data, file, indent=4)
# printing out the results with formatting -f song artist and filename defined above
            print(f"Lyrics for '{song}' by '{artist}' have been written to '{filename}'.")
        else:
            print(f"No lyrics found for '{song}' by '{artist}'.")
    else:
        print(f"Failed to retrieve lyrics for '{song}' by '{artist}'.")

# Test the function with the provided songs
write_lyrics_to_json("Crowder", "Grave Robber", "grave_robber_lyrics.json")
write_lyrics_to_json("Dido", "Thank You", "thank_you_lyrics.json")
write_lyrics_to_json("Rend Collective", "My Lighthouse", "my_lighthouse_lyrics.json")

Lyrics for 'Grave Robber' by 'Crowder' have been written to 'grave_robber_lyrics.json'.
Lyrics for 'Thank You' by 'Dido' have been written to 'thank_you_lyrics.json'.
Failed to retrieve lyrics for 'My Lighthouse' by 'Rend Collective'.


In [27]:
# My lighthouse failed, lets try a new song Firefly. 

def write_lyrics_to_json(artist, song, filename):
    url = f"https://api.lyrics.ovh/v1/{artist}/{song}"
    response = requests.get(url)
    if response.status_code == 200:
        data = response.json()
        if 'lyrics' in data:
            with open(filename, 'w') as file:
                json.dump(data, file, indent=4)
            print(f"Lyrics for '{song}' by '{artist}' have been written to '{filename}'.")
        else:
            print(f"No lyrics found for '{song}' by '{artist}'.")
    else:
        print(f"Failed to retrieve lyrics for '{song}' by '{artist}'.")

# Test the function
write_lyrics_to_json("Ed Sheeran", "Firefly", "Firefly_lyrics.json")

Lyrics for 'Firefly' by 'Ed Sheeran' have been written to 'Firefly_lyrics.json'.


4. Write a function that takes the name of a file that contains song lyrics, loads the file, performs sentiment analysis, and returns the polarity score.  Use this function to print the polarity scores (with the name of the song) of the three files you created in question 3.  Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.

In [37]:
# No need to import anything it has been imported above

def analyze_lyrics(filename):
    # Read the contents of the file
    with open(filename, 'r') as json_file:
        results = json.load(json_file)

    # Check if the JSON data has the key 'lyrics'
    if 'lyrics' in results:
        # Extract the lyrics
        song = results['lyrics']

        # Process the lyrics using spaCy
        doc = nlp(song)

        # Perform sentiment analysis
        sentiment_score = doc._.polarity

        # Print the file name in bold to make it stand out 
        print("\033[1m{}\033[0m ".format(filename))

        # Print the lyrics
        print("\nLyrics of the song:")
        print(song)

        # Print the polarity score of the sentiment analysis
        print("Polarity score of sentiment analysis:", sentiment_score)
        print("__________________________________")  # Adding a  line for separation to show a transition to a new song 
    else:
        print(f"No lyrics found in {filename}.")

# Test the function with multiple files
filenames = ["Firefly_lyrics.json", "grave_robber_lyrics.json", "thank_you_lyrics.json"]
for filename in filenames:
    analyze_lyrics(filename)

Firefly_lyrics.json 

Lyrics of the song:
Paroles de la chanson Firefly par Ed Sheeran
I fell in love next to you
Burning fires in this room
It just fits, light and smooth
Like my feet in my shoes
Little one, lie with me
Sew you heart to my sleeve
We'll stay quiet underneath
Shooting stars, if it helps you sleep

And hold me tight, don't let me breathe
Feeling like you won't believe

There's a firefly loose tonight
Better catch it before it burns this place down
And I lie if I don't feel so right
But the world looks better through your eyes

There's a firefly loose tonight
Better catch it before it burns this place down
And I lie if I don't feel so right
But the world looks better through your eyes

Teach my skin those new tricks
Warm me up with your lips
Heart to heart, melt me down
It's too cold in this town
Close your eyes, lean on me
Face to mouth, lips to cheek
Feeling numb in my feet
You're the one to help me get to sleep

And hold me tight, don't let me breathe
Feeling like you 

I beleive all three songs are positive with the most positive being "Thank You" in my opinion. I would have thought it would be at least .5 as it talked about the best day of my life not sure how you can be more positive than that (Note Score .12) ! "Grave Robber" is back and forth as far a positive and negative goes. It did take the lowest score .096 which I can get behind. "Firefly" is definitly positive as it is a love song with very little that could be interpreted as negative.